### ISSUE TITLE
[Tracking] Spec clarifications

### ISSUE THREAD

**Issue by jcwchen** [onnx/onnx#3651]
Creating this issue to track work items for spec clarifications. There are already multiple GitHub issues regarding op spec not being clear or error in expected outputs. I will collect them in this issue. Feel free to add more work item here. Thanks!

**Comment by jcwchen** [comment]
Old issues which look still valid:

### Pooling related
  - Inconsistent p range between LpNormalization, LpPool and GlobalLpPool: https://github.com/onnx/onnx/issues/1614
  - Ambiguity in Pool spec (AveragePool and MaxPool): https://github.com/onnx/onnx/issues/1922
  - Inconsistencies between MaxPool, AveragePool, and LpPool specs: https://github.com/onnx/onnx/issues/2290
  - MaxUnpool indices incompatible with PyTorch and inconsistent with MaxPool for kernels sizes which aren't a multiple of output size: https://github.com/onnx/onnx/issues/2398
  - operator MaxPool output tensor shape is ambiguously documented: https://github.com/onnx/onnx/issues/2927
  - Confused about AveragePool op ceil_mode attribute: https://github.com/onnx/onnx/issues/3314

### Conv/ConvTranspose related
  - convtranspose with dynamic output shape: https://github.com/onnx/onnx/issues/1772
  - Order of kernel application in Conv & ConvTranspose: https://github.com/onnx/onnx/issues/1886
  - Padding specification unclear for Convolution/AveragePool/MaxPool: https://github.com/onnx/onnx/issues/1651
  - Output shape of Conv not specified: https://github.com/onnx/onnx/issues/1795

### RNN/LSTM related
  - Documentation for RNN, LSTM, GRU shall specify output sequence order when direction is set to reverse: https://github.com/onnx/onnx/issues/626
  - GRU/RNN/LSTM's activation function: https://github.com/onnx/onnx/issues/1166
  - Is RNN hidden_size attribute optional or not? https://github.com/onnx/onnx/issues/1213
  - Spec bug to do with LSTM operator: https://github.com/onnx/onnx/issues/2116

### Other ops
  - [Document] The equivalent pseudo C code of loop: https://github.com/onnx/onnx/issues/3389
  - Rounding behavior of Resize when interpolating on integral types is unspecified: https://github.com/onnx/onnx/issues/3390
  - [ScatterElements] Clarify the requirement on indices uniqueness : https://github.com/onnx/onnx/issues/3234
  - Clarify treatment of Scalars vs 1D tesnors of size 1: https://github.com/onnx/onnx/issues/2428
  - Please update SequenceToSplit spec with default value of 'split': https://github.com/onnx/onnx/issues/2396
  - NonMaxSuppression spatial_dimension should be num_boxes: https://github.com/onnx/onnx/issues/2360
  - The rank for "iou_threshold" and "score_threshold" of NonMaxSuppression is unclear: https://github.com/onnx/onnx/issues/2212
  - Inconsistent ThresholdedRelu spec: https://github.com/onnx/onnx/issues/1983
  - Case sensitivity of operator names and attributes: https://github.com/onnx/onnx/issues/1934
  - Specification for TreeEnsembleClassifier and TreeEnsembleRegressor are inconsistent : https://github.com/onnx/onnx/issues/1925
  - Clarify meaning of operator Constant: https://github.com/onnx/onnx/issues/1799
  - Unsqueeze semantics not described precise enough: https://github.com/onnx/onnx/issues/1751
  - Specifications for SVMClassifier is not precise enough : https://github.com/onnx/onnx/issues/1661
  - Documentation on Gather operator unclear: https://github.com/onnx/onnx/issues/1649
  - Should ReduceL2/ReduceLogSum/ReduceLogSum only support float types? https://github.com/onnx/onnx/issues/1458
  - scale attribute of Crop OP should be marked as optional: https://github.com/onnx/onnx/issues/1211
  - Div operator behavoir for integers: https://github.com/onnx/onnx/issues/1153

**Comment by AlexandreEichenberger** [comment]
Conv operation does not specify if M (channel out) must be a multiple of number of groups. TF Keras and Pytorch have this restriction. #3641

**Comment by kraiskil** [comment]
Difference between Add and Sum could be more explicit, and that difference would be nice to have mentioned in both operator's documentation.
#339 says there was some differences with broadcasting support, but that doesn't seem to be the case anymore.

**Comment by tungld** [comment]
Expected outputs for Round looks incorrect #3755.

**Comment by jcwchen** [comment]
[Spec issue] Rounding or truncation behavior for Cast operator is not specified in the spec #3876

**Comment by gramalingam** [comment]
DequantizeLinear/QuantizeLinear: Spec should be more explicit if the inputs should be 2D or can be N-D, and explain the behavior in N-D case better.

**Comment by gramalingam** [comment]
QuantizeLinear: Spec should be more clear about the precision used for the different operators when using mixed-precision inputs (in computing ```x / y_scale + y_zero_point```).

**Comment by garymm** [comment]
Prelu: The constraint `rank(slope) == rank(X)` OR `slope is unidirectional broadcastable to X` makes sense, but the spec doesn't explicitly say that. That is, it seems to allow `rank(slope) > rank(X)`, which doesn't seem like it makes sense (if it does, could someone please explain?).

https://github.com/onnx/onnx/blob/main/docs/Operators.md#Prelu

**Comment by jcwchen** [comment]
I have a same question as https://github.com/onnx/onnx/issues/2303: "Do subgraph initializer names and input names shadow outer-scope names?" It's better to clarify it in the IR.

**Comment by gramalingam** [comment]
The specification of the "Pad" operator does not describe the intended behavior when there is an interaction between the use of negative pad-values (to truncate/slice) and the use of `reflection` (with a positive pad-value for the opposite side). Please see https://github.com/microsoft/onnxruntime/issues/11828 for a detailed description.

**Comment by AlexandreEichenberger** [comment]
Minor point, but naming of inputs matter in our onnx-mlir code base and we had to write extra templates because A & B for the MatMul, MatMulInteger, and  QLinearMatMul are capitalized differently.

Use `A` & `B`: 
* https://github.com/onnx/onnx/blob/main/docs/Operators.md#MatMul
* https://github.com/onnx/onnx/blob/main/docs/Operators.md#matmulinteger

Use `a` and `b`:
* https://github.com/onnx/onnx/blob/main/docs/Operators.md#QLinearMatMul

I know this may not matter to most, and it is not a big deal, but nevertheless name discrepancy between related ops for which, for example, shape inference can be "commoned" may face minor impediments due to naming conventions.

If this can be fixed without triggering incompatibilities between before/after renaming, then I am for it. If it would trigger incompatibilities, then let us just remember to use names as consistency as possible for future ops.

Thanks

**Comment by gramalingam** [comment]
> If this can be fixed without triggering incompatibilities between before/after renaming, then I am for it. If it would trigger incompatibilities, then let us just remember to use names as consistency as possible for future ops.

My own guess would be that the names should not matter. But, paradoxically, it looks like it would matter for onnx-mlir.

I am in favor of using uniform naming conventions. If anyone has any concerns about changing parameter names of  ops (without changing their version-number), please let us know.

**Comment by AlexandreEichenberger** [comment]
@gramalingam thanks for your answers

Found another case: Conv/ConvTranspose uses `X()` and `W()` but ConvInteger/QLinearConv uses `x()` and `w()`

**Comment by gramalingam** [comment]
Another couple of issues relating to Reduce* ops:

(a) I assume that the Reduce* ops should support the special-case of zero-rank tensors (with a single element). It would be good if the spec explicitly clarifies this.

(b) The spec does not indicate what happens if the input is a tensor with zero elements: e.g., a tensor of shape (100, 0) with reduction along the axis with size 0 (axis 1 in above example). 
(i) The ideal answer for ops like Sum is 0, and Prod is 1. (The identity element for the op.)
(ii) However, for other ops like Max or Min, it can be complicated. An ideal answer for Max() is minus-infinity, but that is valid for only float-like types. For integral-types, it could be min-int.

In any case, some clarification whether this is allowed would be useful.

**Comment by justinchuby** [comment]
> (b) The spec does not indicate what happens if the input is a tensor with zero elements: e.g., a tensor of shape (100, 0) with reduction along the axis with size 0 (axis 1 in above example). (i) The ideal answer for ops like Sum is 0, and Prod is 1. (The identity element for the op.) (ii) However, for other ops like Max or Min, it can be complicated. An ideal answer for Max() is minus-infinity, but that is valid for only float-like types. For integral-types, it could be min-int.

Should we simply return a tensor with zero elements on that axis, so the shape does not change?

**Comment by justinchuby** [comment]
Max and Min can both take rank 0 tensors as inputs but it was not clear from the spec.

**Comment by edgchen1** [comment]
Regarding Pad op spec:

https://github.com/onnx/onnx/blob/0e9debaabfb34cee378ab3ed704c2717e6c168b7/docs/Operators.md?plain=1#L17129-L17142

- should probably also mention optional `axes` input in the first sentence
- "The three supported `modes` are..." and then four are given

**Comment by edgchen1** [comment]
Regarding MeanVarianceNormalization op spec:

https://github.com/onnx/onnx/blob/0f536363edbfdc9653a544c48f69c2c194b38daa/docs/Operators.md?plain=1#L15221-L15226

- How should the default `axis` value be interpreted for inputs with fewer than four dimensions? The "C" reference seems specific to NCHW. This op does not otherwise seem limited to 4D input though.
- Are negative axis values supported?

**Comment by justinchuby** [comment]
https://github.com/onnx/onnx/issues/5493
